crates/dbsp/src/operator/dynamic/group/lag.rs

```diff
@@ -164,207 +157,257 @@ where
-        retractions_factory: &'static dyn Factory<ForwardRetractions<DynPair<I, O>>>,
+        keys_factory: &'static dyn Factory<DynVec<I>>,
         output_val_factory: &'static dyn Factory<O>,
         lag: usize,
         asc: bool,
         project: Box<dyn Fn(Option<&I>, &mut O)>,
         key_cmp: KCF,
-        val_cmp: VCF,
     ) -> Self {
         Self {
             name: format!("{}({lag})", if asc { "lag" } else { "lead" }),
-            retraction_key: output_pair_factory.default_box(),
-            retraction: retraction_factory.default_box(),
             output_pair_factory,
             output_val_factory,
             lag,
             asc,
             project,
-            forward_retractions: retractions_factory.default_box(),
-            backward_retractions: Vec::new(),
+            affected_keys: keys_factory.default_box(),
+            encountered_keys: Vec::new(),
             next_key: 0,
             offset_from_prev: 0,
+            remaining_weight: 0,
             key_cmp,
-            val_cmp,
+            output_pair: output_pair_factory.default_box(),
             _phantom: PhantomData,
         }
     }
 
-    fn push_retraction_from_key(&mut self, key: &I, mut w: ZWeight) {
-        let (i, o) = self.retraction_key.split_mut();
-        key.clone_to(i);
-        (self.project)(None, o);
-
-        self.retraction
-            .from_vals(self.retraction_key.as_mut(), w.erase_mut());
-        self.forward_retractions.push_val(&mut *self.retraction);
-
-        self.backward_retractions.push(BackwardRetraction::new());
+    /// Add `key` to `self.affected_keys`.
+    fn record_affected_key(&mut self, key: &I) -> usize {
+        let idx = self.affected_keys.len();
+        self.affected_keys.push_ref(key);
+        self.encountered_keys.push(EncounteredKey::new());
+        idx
     }
 
-    fn push_retraction_from_ref(&mut self, key: &DynPair<I, O>, w: ZWeight) {
-        self.retraction.from_refs(key, w.erase());
-        self.forward_retractions.push_val(&mut *self.retraction);
+    /// Retract all entries from the output trace whose first element is
+    /// equal to `cursor.key().fst()`; return total weight of removed records.
+    fn retract_key(
+        &mut self,
+        cursor: &mut dyn ZCursor<DynPair<I, O>, DynUnit, ()>,
+        output_cb: &mut dyn FnMut(&mut DynPair<I, O>, &mut DynZWeight),
+    ) -> usize {
+        debug_assert!(cursor.key_valid());
+
+        let mut result = 0;
+
+        let idx = self.record_affected_key(cursor.key().fst());
+
+        while cursor.key_valid() && cursor.key().fst() == self.affected_keys.index(idx) {
+            let weight = **cursor.weight();
+            // The output trace should not contain negative weights by construction.
+            debug_assert!(weight > 0);
+            cursor.key().clone_to(self.output_pair.as_mut());
+            //println!("retract: {:?} with weight {weight}", &self.output_pair);
```

Comment:
If I have so many debugging statements I usually create a local "debug" flag which is statically set to "false" and all the statements are guarded by that flag. Makes it much easier to switch debugging on.

Reply:
that being said, the compiler has an option to set per-Java class tracing levels too, but I don't use that as much.